src/Stratis.Bitcoin.Features.Consensus/ConsensusErrors.cs

```diff
@@ -3,181 +3,177 @@
 
 namespace Stratis.Bitcoin.Features.Consensus
 {
+    /// <summary>
+    /// An exception that is used when consensus breaking errors are found.
+    /// </summary>
     public class ConsensusErrorException : Exception
     {
+        /// <summary>
+        /// Initialize a new instance of <see cref="ConsensusErrorException"/>.
+        /// </summary>
+        /// <param name="error">The error that triggered this exception.</param>
         public ConsensusErrorException(ConsensusError error) : base(error.Message)
         {
             this.ConsensusError = error;
         }
 
+        /// <summary>The error that triggered this exception. </summary>
         public ConsensusError ConsensusError { get; private set; }
     }
 
+    /// <summary>
+    /// A consensus error used to specify different types of reasons a block does not confirm to the consensus rules.
+    /// </summary>
     public class ConsensusError
     {
+        /// <summary>
+        /// The code representing this consensus error.
+        /// </summary>
         public string Code { get; }
 
+        /// <summary>
+        /// A user friendly message to describe this error.
+        /// </summary>
         public string Message { get; }
 
+        /// <summary>
+        /// A method that will throw a <see cref="ConsensusErrorException"/> with the current consensus error.
+        /// </summary>
         public void Throw()
         {
             throw new ConsensusErrorException(this);
         }
 
+        /// <summary>
+        /// Initialize a new instance of <see cref="ConsensusErrorException"/>.
+        /// </summary>
+        /// <param name="code">The error code that represents the current consensus breaking error.</param>
+        /// <param name="message">A user friendly message to describe this error.</param>
         public ConsensusError(string code, string message)
         {
             Guard.NotEmpty(code, nameof(code));
             Guard.NotEmpty(message, nameof(message));
-
+            
             this.Code = code;
             this.Message = message;
         }
 
+        /// <inheritdoc />
         public override bool Equals(object obj)
         {
             ConsensusError item = obj as ConsensusError;
-            if (item == null)
-                return false;
-            return this.Code.Equals(item.Code);
+
+            return (item != null) && (this.Code.Equals(item.Code));
```

Comment:
your `==` and `equals` implementation differs, is that intentional? i.e. is it intentional that you don't have `this.Code == item.Code` here?

Reply:
yes `this.Code == item.Code` will trigger a recursive loop on on the `==` operand just bellow.

Reply:
This code can actually be removed all together we don't need operand overrides really.
But maybe as another PR.

Reply:
i can see the recursive loop here, is Object.ReferenceEquals calling Equals?